File: backend/routes/academic/submission.py
from __future__ import annotations

import uuid
from typing import Optional, List

from fastapi import APIRouter, Depends, status, Request
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from core.database import get_db
from core.utils.response import Response
from core.utils.token import TokenService
from core.middleware.auth import get_token_service, create_auth_dependency, require_lecturer_or_admin, require_student
# StudentAnswerService is used in services and grading; not required at route-level
from services.engine.answer_sheet_extractor import AnswerSheetParser
from services.academic.submission import SubmissionService
from tasks.submission import emit_refresh_dashboard, emit_grade_attempt
from models.academic.question import Question as QuestionModel
from models.academic.exam import Exam as ExamModel
from models.academic.course import Course
from models.account.users import User as UserModel
from models.account.users import UserRole
from models.academic.question import QuestionType
from models.academic.submission import SubmissionStatus
from schemas.academic.submission import (
    SubmissionCreate,
    SubmissionRead,
    SubmissionAttemptRead,
    SubmissionDelete,
)
from schemas.account.users import UserRead

router = APIRouter(prefix="/submissions", tags=["submissions"])




async def _enrich_submission(s, db) -> dict:
    """Add student_name, exam object, course_id, and status to a submission dict."""
    data = s.to_dict()
    # Compute status based on graded_at and latest_score
    if s.graded_at and s.latest_score is not None:
        data['status'] = SubmissionStatus.GRADED.value
    elif s.attempts_count > 0:
        data['status'] = SubmissionStatus.SUBMITTED.value
    else:
        data['status'] = SubmissionStatus.PENDING.value
    # Fetch student name
    student = (await db.execute(select(UserModel).where(UserModel.id == s.student_id))).scalar_one_or_none()
    if student:
        data['student_name'] = student.full_name()
    # Fetch exam and course info
    exam = (await db.execute(select(ExamModel).where(ExamModel.id == s.exam_id))).scalar_one_or_none()
    if exam:
        data['exam'] = exam.to_dict()
        if exam.course_id:
            data['course_id'] = str(exam.course_id)
    return data


class ScanSubmit(BaseModel):
    """Payload for scanning a student's paper answer sheet."""
    exam_id: uuid.UUID
    student_id: uuid.UUID
    pages: List[str]       # base64 images — read by AI parser
    page_urls: List[str]   # storage URLs already uploaded by frontend — stored for audit




def _tenant(user: UserRead) -> Optional[uuid.UUID]:
    return None if user.role in (UserRole.ADMIN, UserRole.SUPERADMIN) else user.tenant_id


# ---------------------------------------------------------------------------
# POST /  — student submits a digital exam attempt
# ---------------------------------------------------------------------------

@router.post("/", status_code=status.HTTP_201_CREATED)
async def submit_exam(
    body: SubmissionCreate,
    request: Request,
    current_user: UserRead = Depends(require_student(get_token_service())),
    db: AsyncSession = Depends(get_db),
):
    service = SubmissionService(db)
    """Submit an exam attempt. Returns immediately; grading runs in background."""
    if not current_user.tenant_id and current_user.role not in (UserRole.ADMIN, UserRole.SUPERADMIN):
        return Response(success=False, error="No tenant assigned to your account", request=request, status_code=status.HTTP_403_FORBIDDEN)

    try:
        # Answers are stored per-question in the student_answers table by the autosave flow.
        # Final submit does not need raw answers supplied in the request body.
        submission, attempt = await service.submit_exam(
            exam_id=body.exam_id,
            student_id=current_user.id,
            tenant_id=_tenant(current_user),
        )
    except ValueError as e:
        return Response(success=False, error=str(e), request=request, status_code=status.HTTP_400_BAD_REQUEST)

    # Grade in background via Kafka worker
    emit_grade_attempt(str(attempt.id), str(body.exam_id))

    # Refresh dashboards in background after submission
    emit_refresh_dashboard(str(current_user.id))
    # Get exam to find lecturer for background refresh
    exam_stmt = select(ExamModel).where(ExamModel.id == body.exam_id)
    exam_result = await service.db.execute(exam_stmt)
    exam = exam_result.scalar_one_or_none()
    # Refresh lecturer dashboard in background if exam has a course with lecturer
    if exam and exam.course_id:
        course_stmt = select(Course).where(Course.id == exam.course_id)
        course_result = await service.db.execute(course_stmt)
        course = course_result.scalar_one_or_none()
        if course and course.lecturer_id:
            emit_refresh_dashboard(str(course.lecturer_id))

    print(f"[log] Student {current_user.id} submitted exam {body.exam_id}, attempt #{attempt.attempt_number}")

    return Response(
        success=True,
        message=f"Attempt #{attempt.attempt_number} submitted — grading in progress",
        data={
            "submission": submission.to_dict(),
            "attempt": attempt.to_dict(),
        },
        request=request,
        status_code=status.HTTP_201_CREATED,
    )


class StartExamRequest(BaseModel):
    exam_id: uuid.UUID


@router.post("/start", status_code=status.HTTP_201_CREATED)
async def start_submission(
    body: StartExamRequest,
    request: Request,
    current_user: UserRead = Depends(require_student(get_token_service())),
    db: AsyncSession = Depends(get_db),
):
    service = SubmissionService(db)
    """Create a Submission record (no attempt) when the student clicks Start Exam."""
    if not current_user.tenant_id and current_user.role not in (UserRole.ADMIN, UserRole.SUPERADMIN):
        return Response(success=False, error="No tenant assigned to your account", request=request, status_code=status.HTTP_403_FORBIDDEN)

    try:
        submission = await service.create_submission(body.exam_id, current_user.id)
    except ValueError as e:
        return Response(success=False, error=str(e), request=request, status_code=status.HTTP_400_BAD_REQUEST)

    return Response(
        success=True,
        message="Submission started",
        data=submission.to_dict(),
        request=request,
        status_code=status.HTTP_201_CREATED,
    )

# ---------------------------------------------------------------------------
# GET /exam/{exam_id}/students  — lecturer sees all students + their submissions
# ---------------------------------------------------------------------------

@router.get("/exam/{exam_id}/students")
async def list_students_with_submissions(
    exam_id: uuid.UUID,
    request: Request,
    page: int = 1,
    per_page: int = 50,
    current_user: UserRead = Depends(require_lecturer_or_admin(get_token_service())),
    db: AsyncSession = Depends(get_db),
):
    service = SubmissionService(db)
    """All students and their full submission + attempt history for a specific exam in the caller's tenant."""
    try:
        offset = (page - 1) * per_page
        results, total = await service.list_students(
            exam_id=exam_id,
            tenant_id=_tenant(current_user),
            limit=per_page,
            offset=offset,
        )
        
        total_pages = (total + per_page - 1) // per_page
        
        return Response(
            success=True,
            message="Students and submissions retrieved",
            data=results,
            page=page, per_page=per_page, total=total,
            request=request,
        )
    except ValueError as e:
        return Response(success=False, error=str(e), request=request, status_code=status.HTTP_404_NOT_FOUND)

# ---------------------------------------------------------------------------
# GET /?exam_id=  — lecturer sees all submissions for an exam
# ---------------------------------------------------------------------------

@router.get("/")
async def list_submissions(
    request: Request,
    exam_id: uuid.UUID,
    status: Optional[str] = None,
    page: int = 1,
    per_page: int = 50,
    current_user: UserRead = Depends(require_lecturer_or_admin(get_token_service())),
    db: AsyncSession = Depends(get_db),
):
    service = SubmissionService(db)
    """List submissions with offset/limit pagination."""
    # Calculate offset from page
    offset = (page - 1) * per_page
    
    # Get submissions
    items = await service.list(exam_id=exam_id, tenant_id=_tenant(current_user), status=status, limit=per_page, offset=offset)
    
    # Get total count for pagination
    total_count = await service.count_for_exam(exam_id=exam_id, tenant_id=_tenant(current_user), status=status)
    
    # Calculate total pages
    total_pages = (total_count + per_page - 1) // per_page
    
    return Response(
        success=True,
        message="Submissions retrieved",
        data=[s.to_dict() for s in items],
        page=page, per_page=per_page, total=total_count,
        request=request,
    )


# ---------------------------------------------------------------------------
# GET /lecturer - lecturer sees all submissions for their exams
# ---------------------------------------------------------------------------

@router.get("/lecturer")
async def list_lecturer_submissions(
    request: Request,
    page: int = 1,
    per_page: int = 50,
    current_user: UserRead = Depends(require_lecturer_or_admin(get_token_service())),
    db: AsyncSession = Depends(get_db),
):
    service = SubmissionService(db)
    """All submissions for exams the lecturer teaches, or all submissions for admins."""
    offset = (page - 1) * per_page
    tenant_id = _tenant(current_user)

    # Admins see all submissions in their tenant; lecturers see only their courses
    if current_user.role in (UserRole.ADMIN, UserRole.SUPERADMIN):
        items = await service.list_for_tenant(tenant_id=tenant_id, limit=per_page, offset=offset)
        total_count = await service.count_for_tenant(tenant_id=tenant_id)
    else:
        items = await service.list_for_lecturer(
            lecturer_id=current_user.id,
            tenant_id=tenant_id,
            limit=per_page,
            offset=offset,
        )
        total_count = await service.count_for_lecturer(
            lecturer_id=current_user.id,
            tenant_id=tenant_id,
        )

    total_pages = (total_count + per_page - 1) // per_page if total_count > 0 else 1

    return Response(
        success=True,
        message="Submissions retrieved",
        data=[await _enrich_submission(s, service.db) for s in items],
        page=page, per_page=per_page, total=total_count,
        request=request,
    )


# ---------------------------------------------------------------------------
# GET /mine?exam_id=  — student sees their own submission + all attempts
# ---------------------------------------------------------------------------

@router.get("/mine")
async def my_submission(
    request: Request,
    exam_id: Optional[uuid.UUID] = None,
    current_user: UserRead = Depends(require_student(get_token_service())),
    db: AsyncSession = Depends(get_db),
):
    service = SubmissionService(db)
    if exam_id:
        # Get submission for specific exam
        submission = await service.get(student_id=current_user.id, exam_id=exam_id)
        if not submission:
            # Return success with null data instead of 404 - student hasn't submitted yet
            return Response(
                success=True, 
                message="No submission found for this exam", 
                data=None,
                request=request,
            )

        attempts = await service.get_attempts(submission.id)
        return Response(
            success=True,
            message="Your submission retrieved",
            data={
                **submission.to_dict(),
                "attempts": [a.to_dict() for a in attempts],
            },
            request=request,
        )
    else:
        # Get all submissions for this student
        submissions = await service.list(student_id=current_user.id, tenant_id=_tenant(current_user), limit=0)
        return Response(
            success=True,
            message="All your submissions retrieved",
            data=submissions,
            request=request,
        )


# Note: draft endpoint removed — autosave writes directly to student_answers.


# ---------------------------------------------------------------------------
# GET /mine/all  — student sees all their submissions (explicit endpoint)
# ---------------------------------------------------------------------------

@router.get("/mine/all")
async def all_my_submissions(
    request: Request,
    current_user: UserRead = Depends(require_student(get_token_service())),
    db: AsyncSession = Depends(get_db),
):
    service = SubmissionService(db)
    """Get all submissions for the current student across all exams."""
    submissions = await service.list(student_id=current_user.id, tenant_id=_tenant(current_user), limit=0)
    return Response(
        success=True,
        message="All your submissions retrieved",
        data=[s.to_dict() for s in submissions],
        request=request,
    )

async def get_attempts(
    submission_id: uuid.UUID,
    request: Request,
    current_user: UserRead = Depends(require_lecturer_or_admin(get_token_service())),
    db: AsyncSession = Depends(get_db),
):
    service = SubmissionService(db)
    submission = await service.get(submission_id, tenant_id=_tenant(current_user))
    if not submission:
        return Response(success=False, error="Submission not found", request=request, status_code=status.HTTP_404_NOT_FOUND)
    attempts = await service.get_attempts(submission_id)
    return Response(success=True, message="Attempts retrieved", data=[a.to_dict() for a in attempts], request=request)


# ... (rest of the code remains the same)

# ---------------------------------------------------------------------------
# POST /scan/  — lecturer scans a student's paper answer sheet
# ---------------------------------------------------------------------------

@router.post("/scan/", status_code=status.HTTP_201_CREATED)
async def scan_answer_sheet(
    body: ScanSubmit,
    request: Request,
    current_user: UserRead = Depends(require_lecturer_or_admin(get_token_service())),
    db: AsyncSession = Depends(get_db),
):
    service = SubmissionService(db)
    """Parse scanned answer sheet pages, map to questions, grade in background."""
    if not body.pages:
        return Response(success=False, error="No pages provided", request=request, status_code=status.HTTP_400_BAD_REQUEST)

    # 1. Load exam questions
    questions = (await db.execute(
        select(QuestionModel).where(QuestionModel.exams.any(ExamModel.id == body.exam_id))
    )).scalars().all()
    if not questions:
        return Response(success=False, error="No questions found for this exam", request=request, status_code=status.HTTP_404_NOT_FOUND)

    # 2. Parse answer sheet pages (fast — just vision extraction, no DB writes)
    question_index = [
        {"number": q.number, "qtype": q.qtype, "options": q.get_options() if q.qtype == QuestionType.MULTIPLE_CHOICE else []}
        for q in questions
    ]
    number_to_id = {q.number: str(q.id) for q in questions}

    try:
        parser = AnswerSheetParser()
        extracted = parser.parse(pages=body.pages, question_index=question_index)
    except RuntimeError as e:
        return Response(success=False, error=str(e), request=request, status_code=status.HTTP_503_SERVICE_UNAVAILABLE)

    # 3. Map question numbers → question IDs
    answers: dict = {}
    for number, answer_value in extracted.items():
        qid = number_to_id.get(str(number))
        if not qid or answer_value is None:
            continue
        q = next((q for q in questions if q.number == str(number)), None)
        if q and q.qtype == "multiple_choice":
            answers[qid] = {"option": str(answer_value).strip().lower()}
        elif q:
            answers[qid] = {"text": str(answer_value).strip()}

    # 4. Save attempt immediately (score=None), enqueue grading
    try:
        submission, attempt = await service.submit_exam(
            exam_id=body.exam_id,
            student_id=body.student_id,
            answers=answers,
            tenant_id=_tenant(current_user),
            scan_pages=body.page_urls,
        )
    except ValueError as e:
        return Response(success=False, error=str(e), request=request, status_code=status.HTTP_400_BAD_REQUEST)

    # Grade scanned attempt in background via Kafka worker
    emit_grade_attempt(str(attempt.id), str(body.exam_id))

    # Refresh dashboards in background after scan submission
    emit_refresh_dashboard(str(body.student_id))
    # Get exam to find lecturer for background refresh
    exam_stmt = select(ExamModel).where(ExamModel.id == body.exam_id)
    exam_result = await db.execute(exam_stmt)
    exam = exam_result.scalar_one_or_none()
    # Refresh lecturer dashboard in background if exam has a course with lecturer
    if exam and exam.course_id:
        course_stmt = select(Course).where(Course.id == exam.course_id)
        course_result = await db.execute(course_stmt)
        course = course_result.scalar_one_or_none()
            if course and course.lecturer_id:
                emit_refresh_dashboard(str(course.lecturer_id))

    return Response(
        success=True,
        message=f"Answer sheet scanned. Attempt #{attempt.attempt_number} — grading in progress",
        data={
            "submission": submission.to_dict(),
            "attempt": attempt.to_dict(),
        },
        request=request,
        status_code=status.HTTP_201_CREATED,
    )

File: frontend/src/apis/submission.ts
import client, { handleEnvelope, handlePaginatedEnvelope, ApiResponse, ApiError } from '@/apis/client'
import { config } from '@/config'
import type { Submission as SharedSubmission, SubmissionAttempt as SharedSubmissionAttempt } from '@/lib/types'

// Use shared types from types.ts
export type Submission = SharedSubmission
export type SubmissionAttempt = SharedSubmissionAttempt

export interface SubmissionAnswer {
  id: string
  attempt_id: string
  question_id: string
  question_text?: string
  answer: string | string[] | number
  is_correct?: boolean
  marks_obtained?: number
  max_marks: number
  feedback?: string
  created_at: string
}

export interface SubmissionCreate {
  exam_id: string
  student_id?: string
  tenant_id?: string
}

export interface SubmissionUpdate {
  status?: 'pending' | 'submitted' | 'graded' | 'returned'
  score?: number
  feedback?: string
}

export interface SubmissionListParams {
  page?: number
  per_page?: number
  search?: string
  status?: Submission['status']
  exam_id?: string
  student_id?: string
  lecturer_id?: string
  tenant_id?: string
  course_id?: string
  start_date?: string
  end_date?: string
}

export interface SubmissionListResponse {
  items: Submission[]
  pagination: {
    page: number
    per_page: number
    total: number
    pages: number
    has_next: boolean
    has_prev: boolean
  }
}

/**
 * List submissions with pagination and filtering
 * Matches backend GET /academic/submissions endpoint
 */
export async function listSubmissions(params?: SubmissionListParams): Promise<SubmissionListResponse> {
  try {
    // For lecturers, use the dedicated endpoint
    if (params?.lecturer_id) {
      const requestParams = {
        page: params?.page || 1,
        per_page: params?.per_page || config.DEFAULT_PAGE_SIZE,
      }

      const response = await client.get<ApiResponse<Submission[]>>('/academic/submissions/lecturer', { 
        params: requestParams 
      })
      
      return handlePaginatedEnvelope<Submission>(response)
    }

    // For students and general use, use the existing endpoint (requires exam_id)
    const requestParams = {
      page: params?.page || 1,
      per_page: params?.per_page || config.DEFAULT_PAGE_SIZE,
      search: params?.search,
      status: params?.status,
      exam_id: params?.exam_id,
      student_id: params?.student_id,
      tenant_id: params?.tenant_id,
      course_id: params?.course_id,
      start_date: params?.start_date,
      end_date: params?.end_date,
    }

    const response = await client.get<ApiResponse<Submission[]>>('/academic/submissions/', { 
      params: requestParams 
    })
    
    return handlePaginatedEnvelope<Submission>(response)
  } catch (error) {
    if (error instanceof ApiError) {
      throw error
    }
    
    if (error instanceof Error) {
      if (error.message.includes('401') || error.message.includes('Not authenticated')) {
        throw new ApiError('Authentication required to view submissions', 401)
      }
      if (error.message.includes('403') || error.message.includes('Forbidden')) {
        throw new ApiError('Access denied to view submissions', 403)
      }
      if (error.message.includes('404') || error.message.includes('Not Found')) {
        throw new ApiError('Submissions endpoint not found', 404)
      }
    }
    
    throw new ApiError('Failed to fetch submissions', 500)
  }
}

/**
 * Get a single submission by ID
 * Matches backend GET /academic/submissions/{submission_id} endpoint
 */
export async function getSubmission(submissionId: string): Promise<Submission> {
  if (!submissionId) {
    throw new ApiError('Submission ID is required', 400)
  }

  try {
    const response = await client.get<ApiResponse<Submission>>(`/academic/submissions/${submissionId}/`)
    return handleEnvelope<Submission>(response)
  } catch (error) {
    if (error instanceof ApiError) {
      throw error
    }
    
    if (error instanceof Error) {
      if (error.message.includes('401') || error.message.includes('Not authenticated')) {
        throw new ApiError('Authentication required to view submission', 401)
      }
      if (error.message.includes('403') || error.message.includes('Forbidden')) {
        throw new ApiError('Access denied to view this submission', 403)
      }
      if (error.message.includes('404') || error.message.includes('Not Found')) {
        throw new ApiError('Submission not found', 404)
      }
    }
    
    throw new ApiError('Failed to fetch submission', 500)
  }
}

/**
 * Create a new submission (start exam)
 * Matches backend POST /academic/submissions endpoint
 */
export async function createSubmission(submissionData: SubmissionCreate): Promise<Submission> {
  if (!submissionData.exam_id) {
    throw new ApiError('Exam ID is required', 400)
  }

  try {
    const response = await client.post<ApiResponse<Submission>>('/academic/submissions/', submissionData)
    return handleEnvelope<Submission>(response)
  } catch (error: any) {
    if (error instanceof ApiError) {
      throw error
    }
    
    // Axios error with response - extract backend message
    if (error.response?.data) {
      const backendError = error.response.data.error || error.response.data.message || 'Failed to start exam'
      throw new ApiError(backendError, error.response.status)
    }
    
    if (error instanceof Error) {
      if (error.message.includes('401') || error.message.includes('Not authenticated')) {
        throw new ApiError('Authentication required to start exam', 401)
      }
      if (error.message.includes('403') || error.message.includes('Forbidden')) {
        throw new ApiError('Access denied to start this exam', 403)
      }
      if (error.message.includes('404') || error.message.includes('Not Found')) {
        throw new ApiError('Exam not found', 404)
      }
    }
    
    throw new ApiError('Failed to start exam', 500)
  }
}

export async function startSubmission(submissionData: { exam_id: string }): Promise<Submission> {
  if (!submissionData.exam_id) {
    throw new ApiError('Exam ID is required to start exam', 400)
  }

  try {
    const response = await client.post<ApiResponse<Submission>>('/academic/submissions/start', submissionData)
    return handleEnvelope<Submission>(response)
  } catch (error) {
    if (error instanceof ApiError) throw error
    throw new ApiError('Failed to start submission', 500)
  }
}

/**
 * Update a submission
 * Matches backend PUT /academic/submissions/{submission_id} endpoint
 */
export async function updateSubmission(submissionId: string, submissionData: SubmissionUpdate): Promise<Submission> {
  if (!submissionId) {
    throw new ApiError('Submission ID is required', 400)
  }

  if (Object.keys(submissionData).length === 0) {
    throw new ApiError('At least one field must be provided for update', 400)
  }

  try {
    const response = await client.put<ApiResponse<Submission>>(`/academic/submissions/${submissionId}/`, submissionData)
    return handleEnvelope<Submission>(response)
  } catch (error) {
    if (error instanceof ApiError) {
      throw error
    }
    
    if (error instanceof Error) {
      if (error.message.includes('401') || error.message.includes('Not authenticated')) {
        throw new ApiError('Authentication required to update submission', 401)
      }
      if (error.message.includes('403') || error.message.includes('Forbidden')) {
        throw new ApiError('Access denied to update this submission', 403)
      }
      if (error.message.includes('404') || error.message.includes('Not Found')) {
        throw new ApiError('Submission not found', 404)
      }
      if (error.message.includes('400') || error.message.includes('Bad Request')) {
        throw new ApiError('Invalid submission data', 400)
      }
    }
    
    throw new ApiError('Failed to update submission', 500)
  }
}

/**
 * Get current user's submission for a specific exam
 * Returns null if no submission exists yet (student hasn't started the exam)
 * Matches backend GET /academic/submissions/mine?exam_id= endpoint
 */
export async function getMySubmission(examId?: string): Promise<Submission | null> {
  try {
    const params = examId ? { exam_id: examId } : {}
    const response = await client.get<ApiResponse<Submission>>('/academic/submissions/mine', { params })
    return handleEnvelope<Submission | null>(response)
  } catch (error) {
    if (error instanceof ApiError) {
      // 404 means no submission yet — return null instead of throwing
      if (error.status === 404) return null
      throw error
    }
    
    if (error instanceof Error) {
      if (error.message.includes('401') || error.message.includes('Not authenticated')) {
        throw new ApiError('Authentication required to view submission', 401)
      }
      if (error.message.includes('403') || error.message.includes('Forbidden')) {
        throw new ApiError('Access denied to view this submission', 403)
      }
      if (error.message.includes('404') || error.message.includes('Not Found')) {
        return null
      }
    }
    
    throw new ApiError('Failed to fetch your submission', 500)
  }
}

/**
 * Get all submissions for the current user across all exams
 * Matches backend GET /academic/submissions/mine/all endpoint
 */
export async function getAllMySubmissions(): Promise<Submission[]> {
  try {
    const response = await client.get<ApiResponse<Submission[]>>('/academic/submissions/mine/all')
    return handleEnvelope<Submission[]>(response)
  } catch (error) {
    if (error instanceof ApiError) {
      throw error
    }
    
    if (error instanceof Error) {
      if (error.message.includes('401') || error.message.includes('Not authenticated')) {
        throw new ApiError('Authentication required to view submissions', 401)
      }
      if (error.message.includes('403') || error.message.includes('Forbidden')) {
        throw new ApiError('Access denied to view submissions', 403)
      }
    }
    
    throw new ApiError('Failed to fetch your submissions', 500)
  }
}

export interface ExamSubmitData {
  exam_id: string
  max_attempts?: number
}

export interface ExamSubmitResult {
  submission: Submission
  attempt: SubmissionAttempt
}

/**
 * Submit exam answers (complete an attempt)
 * Matches backend POST /academic/submissions/submit endpoint
 */
export async function submitExam(submitData: ExamSubmitData): Promise<ExamSubmitResult> {
  if (!submitData.exam_id) {
    throw new ApiError('Exam ID is required', 400)
  }

  try {
    // backend expects POST /academic/submissions/ with minimal payload (exam_id)
    const response = await client.post<ApiResponse<ExamSubmitResult>>('/academic/submissions/', submitData)
    return handleEnvelope<ExamSubmitResult>(response)
  } catch (error) {
    if (error instanceof ApiError) {
      throw error
    }

    if (error instanceof Error) {
      if (error.message.includes('401') || error.message.includes('Not authenticated')) {
        throw new ApiError('Authentication required to submit exam', 401)
      }
      if (error.message.includes('403') || error.message.includes('Forbidden')) {
        throw new ApiError('Access denied to submit this exam', 403)
      }
      if (error.message.includes('404') || error.message.includes('Not Found')) {
        throw new ApiError('Exam or submission not found', 404)
      }
      if (error.message.includes('400') || error.message.includes('Bad Request')) {
        // Pass through the actual backend error message
        const match = error.message.match(/"error":"([^"]+)"/)
        const backendError = match ? match[1] : 'Invalid submission data - check your answers'
        throw new ApiError(backendError, 400)
      }
    }

    throw new ApiError('Failed to submit exam', 500)
  }
}

export default { 
  listSubmissions, 
  getSubmission, 
  createSubmission, 
  updateSubmission,
  getMySubmission,
  getAllMySubmissions,
  submitExam
}
